ACAS-460 LsThingBrowser use flat advanced search results, fix hiding/showing of messages

brianbolt opened the pull request:
## Description
 - Use new "flat" advanced search returns and on click of a row, fetch the full ls thing object
 - Fix issue with dates values rendering as Nan-na-na
 - Fix issues with Search messages not rendering properly
 - Add notification of when max search results has been hit

## Related Issue
ACAS-460

## How Has This Been Tested?
ACAS client tests of new roo route changes
Manual testing of ACASThing browser with 10,000 things, clicking on rows, hitting max limit...etc.

brianbolt commented on modules/Components/src/client/ACASThingBrowser.coffee:
```diff
@@ -69,16 +63,48 @@ class ThingSimpleSearchController extends AbstractFormController
-
+				returnDTO: {
+					thingValues: []
+					thingAttributes: ["id", "codeName"]
```
We need codeName and id for the ls thing editor forms when you click on a row. So we always pass these in the Ajax request.

brianbolt commented on modules/Components/src/client/ACASThingBrowser.coffee:
```diff
@@ -69,16 +63,48 @@ class ThingSimpleSearchController extends AbstractFormController
 			$.ajax
 				type: 'POST'
-				url: "#{@genericSearchUrl}#{@model.get("lsType")}/#{@model.get("lsKind")}?format=nestedfull"
+				url: "#{@genericSearchUrl}#{@model.get("lsType")}/#{@model.get("lsKind")}?format=flat"
```
new flat format request

brianbolt commented on modules/Components/src/client/ACASThingBrowser.coffee:
```diff
@@ -69,16 +63,48 @@ class ThingSimpleSearchController extends AbstractFormController
-				success: (thing) =>
-					@trigger "searchReturned", thing.results
+				success: (response) =>
+					@trigger "searchReturned", response
```
Instead of only passing the "results" we pass the full object so we can use the maxResults and numberOfResults to display a message to the user about hitting the limit of results.

brianbolt commented on modules/Components/src/client/ACASThingBrowser.coffee:
```diff
@@ -163,15 +189,34 @@ class ACASThingBrowserCellController extends Backbone.View
+		if @options.thingModelExample?
+			modelValue = this.options.thingModelExample.get(@configs.key)
+			if modelValue instanceof Value && modelValue.get("lsType") == "dateValue" && !@configs.formatter?
+				content = UtilityFunctions::convertMSToYMDDate(content)
```
There is a lot of magic here, but this is all about determining the type of response (Thing, plain backbone model, or plain javascript key/pair object) and getting their keys.  One of the huge slowdowns of this entire thing was that we were creating a collection of thousands of LsThings which are really slow to instantiate. The default Browser now works using plain javascript objects in the rows (flat response),  but it's possible to pass lsthings of Backbone objects to this class as well (without search browser controller).

brianbolt commented on modules/Components/src/client/ACASThingBrowser.coffee:
```diff
@@ -163,15 +189,34 @@ class ACASThingBrowserCellController extends Backbone.View
+
+		if @options.thingModelExample?
+			modelValue = this.options.thingModelExample.get(@configs.key)
+			if modelValue instanceof Value && modelValue.get("lsType") == "dateValue" && !@configs.formatter?
```
This particular line is important for date values.  We now pass one default plain instantiated ls thing to this method so that we can look up the keys and determine if we should format the returned values as dates.

brianbolt commented on modules/Components/src/client/ACASThingBrowser.coffee:
```diff
@@ -331,42 +380,83 @@ class ACASThingBrowserController extends Backbone.View
+						el: @$('.bv_thingController')
+					@renderController()
+				error: (model, response, options) =>
+					@$('.bv_thingController').html "Error fetching thing"
```
When a row is clicked, if the row is not a Thing instance (it's a plain javascript object by default now), then we use the row (which includes the codeName and id) and fetch it from the server and pass it to the form editor controller.

bffrost commented on modules/Components/src/client/ACASThingBrowserView.html:
```diff
@@ -71,14 +71,15 @@ <h5>Searching for <%-thingName%>s containing text "<i class="bv_exptSearchTerm">
         </div>
-
         <div class="bv_thingTableController hide" style="margin:0px; padding:10px; overflow: auto; overflow: auto;"></div>
+        <label class="bv_maxThingBrowserSearchResultsReached" style="margin-left:10px;"></label>
```
Is this same `bv_maxThingBrowserSearchResultsReached` intended to be in here twice?

brianbolt replied:
Thanks, removed the extra one

bffrost commented on modules/Components/src/client/ACASThingBrowser.coffee:
```diff
@@ -331,42 +380,83 @@ class ACASThingBrowserController extends Backbone.View
-		@thingController = new @controllerClass
-			el: @$('.bv_thingController')
-			model: thing
-			readOnly: true
```
I don't see `readOnly` in your new split implementation below. I think that is important to pass, right?

brianbolt replied:
Thanks.  I did not remove this on purpose. It's strange it's hardcoded to true but I'm not fixing that here.

brianbolt replied:
Fixed by adding it back

bffrost reviewed (approved):
This is pretty complex. I want to give @dalejerikson a chance to review and digest this.

I noticed a few places where we might've left out something important, but otherwise this looks like a nice cleanup.

dalejerikson commented on modules/Components/src/client/ACASThingBrowser.coffee:
```diff
@@ -57,6 +50,7 @@ class ThingSimpleSearchController extends AbstractFormController
 			defaultQueryTerms =
 				queryString: "#{thingSearchTerm}"
 				queryDTO:
+					maxResults: @options.maxResults
```
Is there a "maximum" maxResults limit? Wondering what stress tests look like.

brianbolt replied:
I missed your comment here @dalejerikson . I tasted with 10,0000 and it was fairly performant depending on the needs. I think you'll have to play around with it and find the acceptable limit.

dalejerikson reviewed (approved):
Makes sense to me! Helped to look at https://github.com/mcneilco/acas-roo-server/pull/367 first. Excited about the performance increase! 🥇

bffrost commented:
@dalejerikson Since these changes are really to enable your work, let's hold off on merging this PR into the release until you've gotten a chance to test out this module within your code. That way if you find other small things to tweak, we can fix them on this branch and avoid temporarily breaking anything in the release branch.

dalejerikson commented:
> @dalejerikson Since these changes are really to enable your work, let's hold off on merging this PR into the release until you've gotten a chance to test out this module within your code. That way if you find other small things to tweak, we can fix them on this branch and avoid temporarily breaking anything in the release branch.

Alright sounds good. I'll try to find some time today in between meetings and test this out to observe behavior.

dalejerikson commented:
When implementing this in the context of CompBio I noticed the functionality working well. The performance increase is also really nice. However, this might break some parts of `ACASThingBrowserController`. I noticed the following when running this branch locally... 

Can you confirm if this is true on your end or if this a misconfiguration mistake on my part? 

<img width="541" alt="Screen Shot 2022-10-17 at 12 18 14 PM" src="https://user-images.githubusercontent.com/97194463/196229858-6e808c56-4e17-4555-a1b2-bda25b736794.png">

brianbolt commented:
Thanks @dalejerikson just pushed a fix for this and 2 other issues:

1) Fix for the thing browser controller: This was because thing.attributes didn't exist.  I had made a change at the last minute to pass the plain javascript key/pair object to this method rather than a generic backbone model (non thing backbone model) which would have attributes and I broke the creation of the model.
2) I fixed the Edit button to open the deep link to example things (see change to ControllerRedirectConf.coffee)
3) Fixed the readonly portion of the browser to not lock the example thing when in readonly mode in the browser.

dalejerikson commented:
Thanks @brianbolt , I'll update the CompBio Server, do some testing and see if there's any additional abnormal experiences.

brianbolt commented:
Nothing further from me. Thanks for testing it out and finding some issues.  Merging now.
